plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs

```diff
@@ -730,39 +769,54 @@ compileExpr e = traceCompilation 2 ("Compiling expr:" GHC.<+> GHC.ppr e) $ do
     -- to know we're looking at fromString.
     -- We can safely commit to this match as soon as we've seen fromString - we won't accept
     -- any applications of fromString that aren't creating literals of our builtin types.
-    (strip -> GHC.Var (GHC.idDetails -> GHC.ClassOpId cls)) `GHC.App` GHC.Type ty `GHC.App` _ `GHC.App` content
+    (strip -> GHC.Var (GHC.idDetails -> GHC.ClassOpId cls))
+      `GHC.App` GHC.Type ty `GHC.App` _dict `GHC.App` content
       | GHC.getName cls == GHC.isStringClassName ->
-          case GHC.tyConAppTyCon_maybe ty of
-            Just tc -> case stringExprContent (strip content) of
-              Just bs ->
-                if
-                    | GHC.getName tc == bsTyName -> pure $ PIR.Constant annMayInline $ PLC.someValue bs
-                    | GHC.getName tc == stringTyName -> case TE.decodeUtf8' bs of
-                        Right t -> pure $ PIR.Constant annMayInline $ PLC.someValue t
-                        Left err ->
-                          throwPlain . CompilationError $
-                            "Text literal with invalid UTF-8 content: " <> (T.pack $ show err)
-                    | otherwise ->
-                        throwSd UnsupportedError $
-                          "Use of fromString on type other than builtin strings or bytestrings:" GHC.<+> GHC.ppr ty
-              Nothing ->
-                throwSd CompilationError $
-                  "Use of fromString with inscrutable content:" GHC.<+> GHC.ppr content
+          case GHC.tyConAppTyCon_maybe ty of -- extract Type constructor without arguments
+            Just tc ->
+              if
+                | GHC.getName tc == builtinByteStringTyName ->
+                    case stringExprContent AsBytes (strip content) of
+                      Nothing ->
+                        throwSd CompilationError $
+                          "Use of fromString @BuiltinByteString with inscrutable content:"
+                            GHC.<+> GHC.ppr content
+                      Just bs ->
+                        pure $ PIR.Constant annMayInline $ PLC.someValue bs
+                | GHC.getName tc == stringTyName ->
+                    case stringExprContent AsText (strip content) of
+                      Nothing ->
+                        throwSd CompilationError $
+                          "Use of fromString @BuiltinString with inscrutable content:"
+                            GHC.<+> GHC.ppr content
+                      Just bs ->
+                        case TE.decodeUtf8' bs of
```

Comment:
Hm, `decodeUtf8'` didn't work for bytes, but works for text? What if text _is_ bytes though? Or am I misunderstanding this comment:

> I tried using Data.Text.Encoding.decodeUtf8 and it didn't work: the error was saying about invalid utf-8 bytes. Maybe the implementation in GHC is not compatible with the one in Text (Currently GHC ships with at least five UTF-8 implementations)

?

Reply:
I guess that I don't have unit test with the bytes that break `Text.decodeUtf8`, m.b. it was `\xFF` or `\x00` that `Text` choke on.

Reply:
It was `\xC0`:
```
Cannot decode byte '\xc0': Data.Text.Encoding: Invalid UTF-8 stream
```